### ISSUE TITLE
What is the best practice to make train/eval consistent when batch norm is used?

### ISSUE THREAD

**Issue by xxww** [tensorflow/lingvo#210]
Thanks for the great work on the Lingvo framework. Lately I realize that if I turn on batch norm, the default behavior of Lingvo will cause train/eval to differ.

As a quick measure to fix this issue, I turn on this flag "use_moving_avg_in_training" in BatchNorm layer. And it is defined here: https://github.com/tensorflow/lingvo/blob/master/lingvo/core/bn_layers.py#L65.

However this flag comes with some restrictions, specifically, here: 
https://github.com/tensorflow/lingvo/blob/master/lingvo/core/bn_layers.py#L247

```
      if p.use_moving_avg_in_training:
        beta = 0.0
        gamma = 1.0
      else:
        beta = theta.beta
        gamma = theta.gamma
```

Actually it is quite common for Computer Vision models to apply the moving average mean/var *and* still learn a separate beta and gamma. A prominent example comes from Resnet. It is a well-known trick to initialize the gamma on the skip block at 0 (here it is fixed at 1). This trick alone accounts for about 0.2-0.3% improvement in accuracy, which is quite meaningful.

I am just curious if there is a better way to make Lingvo train/eval consistent with extensive batch norm in CV models?

**Comment by xxww** [comment]
To follow up on this, I am training an Resnet50 model with the standard Imagenet dataset.

And I am getting following numbers after 21 epochs:
Train:
  acc5:0.976, accuracy:0.847, loss:0.56782937
Eval with 10k random samples from validation:
  acc5: 0.521, accuracy: 0.199, loss: 4.091722

The gaps are pretty sizable. Is there someway to make this gap more reasonable?

**Comment by JiahuiYu** [comment]
@xxww Hi we have enabled the BN gamma initialization to zeros as an option. Please checkout this commit: https://github.com/tensorflow/lingvo/commit/67da26789893f49d56877031b01b3fd3cfaf948d

Moreover, please do remember if you want to apply this technique, only enable gamma_zero_init only at the last layer in each res block. Also, just simply keep p.use_moving_avg_in_training to False. It has nothing to do with gamma_zero_init technique you mentioned.

**Comment by xxww** [comment]
Thanks for the great work! I now disabled use_moving_avg_in_training. But I still see very big train/eval gap. After 100 epochs on Imagenet with Resnet50, I got this:

Train: acc5:0.94921875 accuracy:0.8046875 loss:0.74647748
Eval: acc5: 0.89494085, accuracy: 0.6844629, loss: 1.2805895

The test accuracy is lowered by 12%. :(

**Comment by JiahuiYu** [comment]
@xxww Did you follow other techniques like warmup learning rate schedule as introduced in https://arxiv.org/abs/1706.02677?

**Comment by xxww** [comment]
Yeah, I have warm-up etc. But the train/eval diff is just too much. To be specific:

I am using batch_size = 256, training_epochs = 100, warmup_epochs=1, lr = 1e-3, weight_decay = 1e-5. I am using linear rampup and exponential decay to 1e-3 * lr.

I am also experimenting another run with batch_size=1024, warmup_epochs=3, lr=2e-3 and exponential decay to 1e-2 * lr. But so far, it still has a similar train/eval diff.
